fix: codeigniter in a subdirectory + indexPage results in 404

Franky5831 opened the pull request:
**Description**
This PR fixes the CodeIgniter routing for projects that run in **subdirectories** and have a **indexPage**.

If you run the framework in the subdirectory of a webserver:
```
/var/www/html/index.html
/var/www/html/your-codeigniter-app
/var/www/html/some-other-app
```
resulting in the following url: `http://localhost:8080/your-codeigniter-app`

And use a indexPage (the `$indexPage` value can be configured in `\Config\App`) whatever route you try to get will result in 404.
`http://localhost:8080/sub-directory/index.php/your-route` will be routed as `/index.php/your-route`, not as `/your-route`

After performing the change I tested it on multiple apps I developed on the CodeIgniter4 framework both on apache and nginx, trying to run them with combinations of `indexPage` configs, inside/outside subdirectories (even multiple levels deep), everything seems to be working fine. I also ran the PHPUnit tests (excluding _DatabaseLive_ and _CacheLive_), all tests passed.

Problem reported here: https://github.com/codeigniter4/CodeIgniter4/issues/9602#issuecomment-2996700517

**Checklist:**
- [x] Securely signed commits
- [ ] Component(s) with PHPDoc blocks, only if necessary or adds value
- [x] Unit testing, with >80% coverage
- [ ] User guide updated
- [x] Conforms to style guide

Franky5831 commented:
Hey, I don't understand why this action failed: https://github.com/codeigniter4/CodeIgniter4/actions/runs/15828226403/job/44613949874?pr=9607

Can someone help me figure it out?
Thanks.

michalsn commented on system/HTTP/SiteURIFactory.php:
```diff
@@ -111,8 +111,40 @@ public function detectRoutePath(string $protocol = ''): string
      */
     private function parseRequestURI(): string
     {
+        $appConfig = config(App::class);
```
Please use `$this->appConfig` instead.

Franky5831 replied:
Just updated

Franky5831 replied:
I just noticed that if I use `$this->appConfig` I am not able to change the value of it during tests.
In my test I wrote:
```php
config(App::class)->baseURL = $standardUrl . $option["subDir"];
```
if I do that and then use `$this->appConfig` it does not get the updated value set by the test.

Do you know how I can do that?
I can always go deeper and try to figure that out my self.

michalsn replied:
In your tests, you can create a new instance of the config - `$config = new App()`, make changes, and then use it with `createSiteURIFactory($config)` - that should work.

Franky5831 replied:
@michalsn you were right, I just updated

michalsn reviewed (changes requested):
Could you please add some tests to confirm that the scenario you're trying to fix is now working?

I'm also not entirely sure about the implementation you've chosen. I hadn't noticed before that we have access to `App::indexPage` here… are you sure we can't handle this more simply? For example, by just checking whether `index.php` should be stripped from `$path`?

michalsn commented:
> We could try to remove the first occurrence of the indexPage from the uri before using it for the routing. I'm pretty sure that would work, what do you think?

I was thinking about something like this:

```php
// Cleanup: if indexPage is still visible in the path, remove it
if ($this->appConfig->indexPage !== '' && str_starts_with($path, $this->appConfig->indexPage)) {
    $remainingPath = substr($path, strlen($this->appConfig->indexPage));
    // Only remove if followed by '/' (route) or nothing (root)
    if ($remainingPath === '' || str_starts_with($remainingPath, '/')) {
        $path = ltrim($remainingPath, '/');
    }
}
```

Added before this block of code: https://github.com/codeigniter4/CodeIgniter4/blob/develop/system/HTTP/SiteURIFactory.php#L149

---

Please update your branch to get all changes added in the meantime: https://github.com/codeigniter4/CodeIgniter4/blob/develop/contributing/workflow.md#updating-your-branch

Franky5831 commented:
> > We could try to remove the first occurrence of the indexPage from the uri before using it for the routing. I'm pretty sure that would work, what do you think?
> 
> I was thinking about something like this:
> 
> ```
> // Cleanup: if indexPage is still visible in the path, remove it
> if ($this->appConfig->indexPage !== '' && str_starts_with($path, $this->appConfig->indexPage)) {
>     $remainingPath = substr($path, strlen($this->appConfig->indexPage));
>     // Only remove if followed by '/' (route) or nothing (root)
>     if ($remainingPath === '' || str_starts_with($remainingPath, '/')) {
>         $path = ltrim($remainingPath, '/');
>     }
> }
> ```

This won't work, you can now test it yourself using the test I created, `str_starts_with($path, $this->appConfig->indexPage)` won't work because it will start with the path to the subdirectory; if we have `localhost:8080/directory/index.php` it will not remove index.php as the uri starts with `directory`.

I can refactor my code to keep it in line with what you wrote.

TimexPeachtree commented:
I am late to review how this will affect, as currently we do .htaccess and config.php, etc.,  workarounds to make it work in subdirectories.

We can use/modify the same htaccess rules in IIS (Webserver in Windows) using URL Rerouting. Bear with me windows dev here. 

I want to know will we be needing to further modified .htaccess rules to make use of this changes or is this changes makes it redundant to use those workarounds.

If this changeset makes it redundant, then should we replace .htaccess same as release provided .htaccess and just make this config changes to allow CodeIgniter to be used from subdirectories.

Please feel free to Correct me if I have not got it right.

michalsn commented:
@Franky5831 Your test isn't valid as it produces incorrect values for superglobals when `indexPage` is empty.
Also, I don’t see it covering the scenario you’re aiming to handle. This isn’t a rewrite test - it's testing serving the app from subfolders *without* rewrite, which hasn’t caused any issues.

Could you also try using a `DataProvider` for testing? There are examples in the same file, and it would help make the tests more readable and maintainable.

IMO, the proposed version would work - I might try to test it more this evening.

Franky5831 commented:
> @Franky5831 Your test isn't valid as it produces incorrect values for superglobals when `indexPage` is empty.

I just fixed it.

> Also, I don’t see it covering the scenario you’re aiming to handle. This isn’t a rewrite test - it's testing serving the app from subfolders _without_ rewrite, which hasn’t caused any issues.

I created the `$subDirectoryAppsOptions` array where I listed all the cases with the resulting `$_SERVER['SCRIPT_NAME']` being hard coded. I could not find a way to run a webserver instance inside the tests and get the `$_SERVER['SCRIPT_NAME']` from there.

> Could you also try using a `DataProvider` for testing? There are examples in the same file, and it would help make the tests more readable and maintainable.

I just added that.
